packages/form-js-playground/src/components/PlaygroundRoot.js

```diff
@@ -193,59 +165,77 @@ export function PlaygroundRoot(props) {
       }
     });
 
-    const formContainer = formContainerRef.current;
-    const editorContainer = editorContainerRef.current;
-    const dataContainer = dataContainerRef.current;
-    const resultContainer = resultContainerRef.current;
-
-    dataEditor.attachTo(dataContainer);
-    resultView.attachTo(resultContainer);
-    form.attachTo(formContainer);
-    formEditor.attachTo(editorContainer);
+    inputDataEditor.attachTo(inputDataContainerRef.current);
+    outputDataEditor.attachTo(outputDataContainerRef.current);
 
     return () => {
-      dataEditor.destroy();
-      resultView.destroy();
-      form.destroy();
+      inputDataEditor.destroy();
+      outputDataEditor.destroy();
+      formViewer.destroy();
       formEditor.destroy();
     };
-  }, []);
+  }, [ additionalModules, editorAdditionalModules, editorProperties, emit, exporterConfig, propertiesPanelConfig, viewerAdditionalModules, viewerProperties ]);
 
+  // initialize data through props
   useEffect(() => {
-    dataEditorRef.current.setValue(toString(initialData));
-  }, [ initialData ]);
+    if (!config.initialSchema) {
+      return;
+    }
+
+    load(config.initialSchema, config.initialData || {});
+  }, [ config.initialSchema, config.initialData, load ]);
```

Comment:
I would have these be the default values of the `useState` calls.

```js
const [ schema, setSchema ] = useState(config.initialSchema);
const [ data, setData ] = useState(config.initialData || {});

useEffect(() => {
  if (!schema) { 
    return; 
  }    
  formEditorRef.current.importSchema(schema, data);
  inputDataRef.current.setValue(toString(data));
  if (inputDataContainerRef.current) {
    const variables = getSchemaVariables(schema);
    inputDataRef.current.setVariables(variables);
  }
}, [ data, schema, getSchemaVariables ]);
```

This means you can git rid of the effect on 188-190, 192-197 and all `load` does it set the states.